frontend/src/components/Scratch/hooks/useLanguageServer.ts

```diff
@@ -83,11 +97,30 @@ export default function useLanguageServer(scratch: api.Scratch, sourceEditor: Mu
         sourceEditor.current?.dispatch({ effects: StateEffect.appendConfig.of(sourceLsExtension) })
         contextEditor.current?.dispatch({ effects: StateEffect.appendConfig.of(contextLsExtension) })
 
+        setSaveSource(() => _saveSource)
+        setSaveContext(() => _saveContext)
+
         return () => {
             (async () => {
                 await _lsClient.exit()
             })()
         }
 
-    }, [ClangdStdioTransportModule, languageId, contextEditor, sourceEditor])
+    }, [initialScratchState, ClangdStdioTransportModule, languageId, contextEditor, sourceEditor])
+
+    const saveSourceRet = () => {
+        (async () => {
+            if (saveSource)
+                await saveSource(scratch.source_code)
+        })()
+    }
+
+    const saveContextRet = () => {
+        (async () => {
+            if (saveContext)
+                await saveContext(scratch.context)
+        })()
```

Comment:
This will have the same behaviour as just calling `saveContext(scratch.context)` and discarding the Promise, no?

Reply:
Oops; you're right. I think that's the result of me refactoring out some more complicated behavior, and then forgetting to unwrap that.